### ISSUE TITLE
HMR taking much longer with webpack@5 and next@canary

### ISSUE THREAD

**Issue by PepijnSenders** [vercel/next.js#20760]
**What version of Next.js are you using?**

10.0.5-canary.9

**What version of Node.js are you using?**

v15.3.0

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**How are you deploying your application?**

next start

**Describe the Bug**

In dev mode with webpack@5 HMR takes considerably longer than with webpack@4 for the same change and exactly the same codebase.

**Expected Behavior**

HMR being faster or at the same speed.

**To Reproduce**

Still need to isolate the issue, which is really hard because this page is around 700kB total!

The webpack@4 baseline uses the following versions:

```
├─ next@npm:10.0.3
│  └─ webpack@npm:4.44.1 (via npm:4.44.1)
```

The baseline refresh time of HMR is around a second:

<img width="354" alt="Screen Shot 2021-01-04 at 8 07 55 PM" src="https://user-images.githubusercontent.com/2052659/103595140-b6122400-4ec8-11eb-85cd-83b80b69d0db.png">

Compared with doing the same action with webpack@5:

```
├─ next@npm:10.0.5-canary.9
│  └─ webpack@npm:5.11.1 (via npm:5.11.1)
```

<img width="375" alt="Screen Shot 2021-01-04 at 8 10 53 PM" src="https://user-images.githubusercontent.com/2052659/103595248-f7a2cf00-4ec8-11eb-89bf-ceb5f05a3921.png">

This takes around 12 seconds (vs. 1 second with webpack@4) and there's an extra event as well:

```json
{
	"event": "serverOnlyChanges",
	"pages": ["/legacy/plans"]
}
```

and the built event has some errors regarding anonymous arrow functions that didn't have anything to do with the change.

We initially thought this issue was related to the context issue #18917, but it's not fixed on the latest canary.

**Comment by Timer** [comment]
We need a full reproduction before we can take a look at this!

**Comment by PepijnSenders** [comment]
@Timer I just tested `config.cache.type = 'memory';` which makes it much faster:

<img width="668" alt="Screen Shot 2021-01-05 at 9 06 35 AM" src="https://user-images.githubusercontent.com/2052659/103655514-61f15900-4f35-11eb-8839-1ee4e1387aca.png">

If you have any recommendations in the meantime on how to isolate this issue, please let me know!

**Comment by PepijnSenders** [comment]
Was hoping the [speed-measure-webpack-plugin](https://github.com/stephencookdev/speed-measure-webpack-plugin) would give some more insights, but it doesn't show any anomalies in the plugins/loaders that we use either:

Webpack 4:
```
 SMP  ⏱  
General output time took 0.143 secs

 SMP  ⏱  Plugins
WellKnownErrorsPlugin took 0.065 secs
NextJsRequireCacheHotReloader took 0.005 secs
DefinePlugin took 0.001 secs
ChunkNamesPlugin took 0 secs
NextJsSsrImportPlugin took 0 secs
PagesManifestPlugin took 0 secs

 SMP  ⏱  
General output time took 0.169 secs

 SMP  ⏱  Plugins
NextJsRequireCacheHotReloader took 0.006 secs
DefinePlugin took 0.001 secs
ReactFreshWebpackPlugin took 0 secs
ChunkNamesPlugin took 0 secs
DropClientPage took 0 secs
HotModuleReplacementPlugin took 0 secs
WellKnownErrorsPlugin took 0 secs
ReactLoadablePlugin took 0 secs
BuildManifestPlugin took 0 secs
```

Webpack 5 (cache.type = 'memory'):

```
 SMP  ⏱  
General output time took 8.28 secs

 SMP  ⏱  Plugins
WellKnownErrorsPlugin took 0.001 secs
PagesManifestPlugin took 0.001 secs
DefinePlugin took 0 secs
NextJsSsrImportPlugin took 0 secs
NextJsRequireCacheHotReloader took 0 secs

 SMP  ⏱  Loaders
next took 0.177 secs
  module count = 1





 SMP  ⏱  
General output time took 8.75 secs

 SMP  ⏱  Plugins
ReactLoadablePlugin took 0.008 secs
BuildManifestPlugin took 0.002 secs
NextJsRequireCacheHotReloader took 0.002 secs
ReactFreshWebpackPlugin took 0.001 secs
HotModuleReplacementPlugin took 0.001 secs
WellKnownErrorsPlugin took 0.001 secs
ProvidePlugin took 0 secs
DefinePlugin took 0 secs
DropClientPage took 0 secs

 SMP  ⏱  Loaders
@next/react-refresh-utils, and 
next took 0.382 secs
  module count = 1
```

Webpack 5 (cache.type = 'fs'):

```
 SMP  ⏱  
General output time took 8.46 secs

 SMP  ⏱  Plugins
NextJsRequireCacheHotReloader took 0.001 secs
DefinePlugin took 0 secs
NextJsSsrImportPlugin took 0 secs
WellKnownErrorsPlugin took 0 secs
PagesManifestPlugin took 0 secs

 SMP  ⏱  Loaders
next took 0.123 secs
  module count = 1

 SMP  ⏱  
General output time took 9.19 secs

 SMP  ⏱  Plugins
ReactLoadablePlugin took 0.006 secs
ProvidePlugin took 0.001 secs
WellKnownErrorsPlugin took 0.001 secs
BuildManifestPlugin took 0.001 secs
ReactFreshWebpackPlugin took 0 secs
DefinePlugin took 0 secs
DropClientPage took 0 secs
HotModuleReplacementPlugin took 0 secs
NextJsRequireCacheHotReloader took 0 secs

 SMP  ⏱  Loaders
@next/react-refresh-utils, and 
next took 0.432 secs
  module count = 1
```

**Comment by Timer** [comment]
I just tried switching from filesystem to memory and it did not yield a meaningful speedup. Can you please share your project so we can trace the slowdown?

**Comment by PepijnSenders** [comment]
It's not a public project, and yea the switch also didn't yield a meaningful speed difference when you look at the stats in https://github.com/vercel/next.js/issues/20760#issuecomment-754745877.

What kind of project are you testing this on? I'm afraid it might only show up when you have a significant amount of code that needs to be compiled.

**Comment by Timer** [comment]
We're testing it on vercel.com, so a large project!

**Comment by PepijnSenders** [comment]
I believe it, I will try some more and see if removing certain parts isolate the issue. When you try it on vercel.com there's no change at all between webpack 5 and 4 refresh times?

**Comment by PepijnSenders** [comment]
Some more after hour debugging:

If the initial page that you compile is a smaller page, the compile and refresh time is very fast, like with webpack@4. If you compile a bigger and slower page after and then go back to smaller page the smaller page is suddenly very slow to compile as well.

Going to see if I can find anything that gets stuck in the memory heap and if this is the same for both cache types.

**Comment by Timer** [comment]
On vercel.com webpack 5 seems a bit faster than webpack 4 for us.

**Comment by PepijnSenders** [comment]
> This is a good observation. Could you define the size of a "big" page? Are you using a lot of app-code, or very heavy npm dependencies?

This is from the output:

```
├ λ /legacy/plans                              68.9 kB         509 kB
```

It's a lot of app code, lots of dynamic imports, and experiments.

**Comment by PepijnSenders** [comment]
I will check if I can see anything in heap snapshots tomorrow. It might be interesting to compare a heap snapshot of when a smaller page is still "fast", and then after going to a bigger page and it being very slow after.

**Comment by Timer** [comment]
509 kB is pretty large... it might be due to a ton of dynamic imports. Can you see if this reproduces on a "large page" that doesn't use `import()` or `next/dynamic`?

**Comment by PepijnSenders** [comment]
@Timer Another find, when I don't clear the `.next/cache` the small pages will also stay slow.

**Comment by PepijnSenders** [comment]
Also from my investigation on the heap snapshots, it seems like a lot of timers are set but never removed: https://github.com/webpack/webpack/blob/master/lib/cache/IdleFileCachePlugin.js#L140. Will need to check some more if that's actually the case or if there a just a lot of them.

**Comment by Timer** [comment]
Could you elaborate on "don't clear" @PepijnSenders? You should **never** be clearing `.next/cache`.

**Comment by PepijnSenders** [comment]
@Timer sorry missed your response, I know you should never clear it but my hunch is that the new webpack is adding a lot unnecessary things to the cache making everything slower. So as a test I removed the `.next/cache` folder between runs and when I start with a clean project it's much faster (if I don't go to bigger pages). Thought that might be a hint in the right direction.

I also see that the ticket was added to iteration 16, do you need more from me?

**Comment by Timer** [comment]
Yes (still need more info), it was only added to the iteration for investigation, but we're blocked on a full reproduction.

Can you make a mock app with large pages that reproduces this?

**Comment by PepijnSenders** [comment]
@Timer Do you jump on calls with people maybe we can dive into it together? I think creating a mock repo would be a lot of work, which I can do if you don't do that.

**Comment by Timer** [comment]
Unfortunately a lot of our profiling is done with tools that wouldn't make it easy to do over a call. A mock repo would be the best.

**Comment by Timer** [comment]
Can you share your `next.config.js`?

**Comment by PepijnSenders** [comment]
> Can you share your `next.config.js`?

That's complicated, I'll try to create a mock repo soon.

**Comment by cullylarson** [comment]
@PepijnSenders Are you by chance using https://github.com/giuseppeg/styled-jsx-plugin-postcss in your project? If so, I think I know why it might be slow.

**Comment by PepijnSenders** [comment]
@cullylarson thanks for chiming in, but no we don't use any custom plugins.

**Comment by Timer** [comment]
@PepijnSenders could you please try the latest canary?

**Comment by Timer** [comment]
Was hoping this helped: https://github.com/vercel/next.js/pull/21205

**Comment by Timer** [comment]
Any luck on a reproduction?

**Comment by PepijnSenders** [comment]
Haven't had any time to look at that yet. I'll try to get that by the end of this week.

**Comment by PepijnSenders** [comment]
@Timer got a bit creative and created this repo: https://github.com/PepijnSenders/webpack-5-issues-repro

With webpack@4:

<img width="889" alt="Screen Shot 2021-01-26 at 2 47 21 PM" src="https://user-images.githubusercontent.com/2052659/105900557-36b8d180-5fea-11eb-892d-409a4c2236d3.png">

With webpack@5:

<img width="593" alt="Screen Shot 2021-01-26 at 3 19 37 PM" src="https://user-images.githubusercontent.com/2052659/105900534-2ef92d00-5fea-11eb-96d9-274dddb19fb9.png">

The difference is quite palpable.

I added instructions on how I run those tests in the README! Let me know if that helps at all.

**Comment by Timer** [comment]
I've just tested this on the latest Next.js canary and both feel about identical. Can you please try on your real project?

Note to enable webpack 5 you must now do:
```tsx
module.exports = { future: { webpack5: true } };
```

**Comment by PepijnSenders** [comment]
Thanks @Timer, I sadly had no time to make better repro's but I'll test this out today 💪

**Comment by PepijnSenders** [comment]
By the way @Timer, we can't upgrade anyways unless this issue was also resolved: https://github.com/vercel/next.js/issues/21642. I will test the webpack 5 stuff locally first though!

**Comment by Timer** [comment]
We'll take a look at that issue!

**Comment by PepijnSenders** [comment]
Under one second with the latest canary @Timer ! Seems like it works, I will let some devs play around with it as well and get back to you tomorrow.

**Comment by PepijnSenders** [comment]
@Timer , actually one thing. I forgot to enable the .future in `next.config.js` when I enable it all my custom aliases suddenly don't work anymore.

**Comment by PepijnSenders** [comment]
@Timer any idea why the aliases would fail with that setting? I was just using resolutions which looks like would have a similar effect.

**Comment by cullylarson** [comment]
@PepijnSenders It reminds me of an issue I had with another package (https://github.com/ilearnio/module-alias/issues/59). When I used native ES modules, aliases stopped working. I wonder if webpack 5 has a similar issue.

**Comment by PepijnSenders** [comment]
Interesting but I don't think it's specifically webpack 5 in this case, because when I force it to webpack 5 it works. When I set .future.webpack5 to true though in `next.config.js` the aliases are suddenly not defined anymore.

**Comment by cullylarson** [comment]
Doesn't `.future.webpack5` force weback 5?

**Comment by PepijnSenders** [comment]
It does and it also does a couple of other things apparently which breaks resolve.alias in the webpack config.

**Comment by cullylarson** [comment]
Oh, I see what you mean. When you say "force it to webpack 5", you mean force webpack 5 without using the `.future.webpack5` flag.

I wonder if `.future.webpack5` also turns on native ES modules.

**Comment by bfunc** [comment]
I have an issue with auto update page. It stopped working (no updating triangle) on versions of next,js larger than 10. Even default installation from homepage does not not work. Constantly getting error _(canceled) | eventsource_ on __next/webpack-hmr?page=/_ http request. Tried version 10.0.7 on two diffrent machines. My luck that version 10.0.0 works ok

**Comment by PepijnSenders** [comment]
Will open a new issue for this one: https://github.com/PepijnSenders/nextjs-aliases-issue

**Comment by vvo** [comment]
For what it's worth: I too experienced slower rebuild times in dev using webpack 5. For example on an index.js page of 1000 lines with ~20 imports (internal/external modules).

After breaking up the page into different components (multiple files, example: components/Home/Pricing components/Home/Testimonials), with each component being ~100 lines then the rebuild speed is now almost instant when I change Testimonials (vs ~5 seconds before: the time it takes to rebuild the whole index.js page).

tl;dr; rebuilding a complete landing page of 1000 lines will be slow: break it down!

**Comment by timneutkens** [comment]
@vvo are you using next-transpile-modules?

**Comment by vvo** [comment]
@timneutkens yes, for one module indeed, could it be the reason?

**Comment by timneutkens** [comment]
Yes for sure: https://github.com/martpie/next-transpile-modules/pull/179

**Comment by vvo** [comment]
Thanks, @timneutkens, does the "break pages into external file components" advice still make sense you think? 

Just removed next transpile module (not used on the page I am modifying) and on the big page it's better but still slower than updating a component part of the page.

Also, not on an M1 so I guess it could be even faster :D

**Comment by timneutkens** [comment]
> does the "break pages into external file components" advice still make sense you think?

Depends, in general the difference between webpack 4/5 shouldn't be much. Breaking up the components makes things faster as there's less code to be parsed/compiled for a single line change update.

**Comment by timneutkens** [comment]
Hey @PepijnSenders, can this one be closed now, afaik all issues around this were solved.

**Comment by PepijnSenders** [comment]
All solved and working great. Thanks for the good work!

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
